### ISSUE TITLE
Queries on Many-To-Many Relations are always slow

### ISSUE THREAD

**Issue by halvini** [parse-community/Parse-SDK-iOS-OSX#769]
My app is a messenger and I need to refresh all the updated chats of a user when he opens "My Chats" section. Total there are 5M chats in database.

I have following tables:
_User 
Dialogues (or "Chats"; this table has a many-to-many field "users", which includes all the users, who are the members of the Chat; a lot of Chats are group chats, so I couldn`t find a more efficient way to store them; arrays, one-to-many relations and Join tables didn`t work for many specific requests).

To get the latest chats of a user I make the following request:

``` swift
        let chatQuery = PFQuery(className:"Dialogues")
        chatQuery.whereKey("users", equalTo: PFUser.currentUser())
         // Getting results which were updated later than last fetch date
        chatQuery.whereKey("updatedAt", greaterThan: lastFetchDate)     
        chatQuery.findObjectsInBackgroundWithBlock
```

So, this simple query which usually return only few dozen results always is Slow. It`s median (ms) is 10481.933. So it spends 10+ seconds just to get these results.

What`s wrong with that query?

**Comment by nlutsenko** [comment]
Hey @halvini, since you have 5M chats - this would probably lead to a full table scan, since you are querying by the constraint on the array key.

To improve the query performance here - I would recommend building a join table "DialogueUsers" with the following schema:
```
 DialogueUsers:
 "user" : "_User",
 "dialog" : "Dialog", 
 "updatedAt" : "...",
 "activity" : 0,
 ...
```

This would basically designate which users are in which dialog, allowing you to perform a fast comparison query on the Dialogues that the currentUser is in.
So your query would looks something like:
```swift
let query = PFQuery(className: "DialogueUsers")
query.whereKey("user", equalTo: PFUser.currentUser())
query.whereKey("updatedAt", greaterThan: lastFetchDate)     
query.includeKey("dialogue")
query.findObjectsInBackgroundWithBlock
```
There are going to be other changes required:
- When a user joins a dialogue - creating a new `DialogueUsers` object.
- When a new message is posted into the dialogue - increment the activity key on all the `DialogUsers` objects in the current chat (this will also update the `updatedAt` date).

To improve further - you might want to not fetch all of the `DialogueUsers` every time to trigger the update for chats, but rather use Push Notifications to trigger updates for a specific dialogue, and then fetch that dialogue only. Meaning that you need to perform this query only on app start.

Hope this helps, let me know if there is anything else I could help you with.

**Comment by halvini** [comment]
Hey, @nlutsenko, Join Table - is actually what I tried to do earlier and I pm`ed you about that, if you remember. 

The fact is that doesn`t work when we are talking about chats with a number of users more than 2. How do I update "DialogueUsers" schema if there are 100 users in a group chat and they are posting there 1 message per second? It will lead to 100 new requests per second, just to let these users know that a message was written in a chat.

But what if we have like 100 of chats like that one? Or 1000?

I cant update the DialogueUsers all the time. So I tried to create them only once and check if "Dialogue" is updated in the "dialog" Pointer. But the subquery also didn`t work, because it has a limit of 1000. So I ended up creating Many-to-Many Relations.

The fact is that scheme at least works. Even very slow.

So the question is, how do I make it faster by NOT making additional hundreds of queries per second or losing data?

**Comment by halvini** [comment]
Thank you for reopening the issue, @nlutsenko 

Unfortunately, I couldn't reopen it myself. I can't find that button here.

**Comment by nlutsenko** [comment]
After giving it a second thought:
- You can realistically completely skip the optimization on `updatedAt` or `activity` from the last piece - which means, you don't need to perform a big write for every new message, but simply update the dialogue.
- Your query still goes against `DialogueUsers`.
- Since you expanded a pointer via `includeKey:@"dialogue"` - you will also fetch all the dialogues.
- On the client - you can simply sort the dialogues by their `updatedAt`.
- All of this works great, if fetching your `DialogueUsers` is not performed every minute, but rather is limited to 5 minutes or is triggered via push notification.

The way I am thinking here is that you can optimize the query performance by pre-caching the results with a join table, and slightly rework the optimization to still maintain responsiveness on the client.
Let me know if this makes sense.

**Comment by halvini** [comment]
@nlutsenko, I've tried that scenario already too. Well, I wouldn't open an issue if I couldn't solve it on my own. I'm sure you have a lot of work to do, and I hope this experience will help your future clients with similar load.

I.
So, what happens if we follow the scenario described above. My app is a place where a lot of people meet, they open new chats, they create group chats by interests, and average client spends 60 minutes in the app per day. That doesn't mean that most clients do that. On the opposite, that means, several % of the audience spend there whole day, but other clients just visit for 1-10 minutes and leave. In fact, I see more and more people who spend days and nights there, so these active clients have joined thousands chats each already. I'm not dreaming. I'm personally connected to more than 10k chats. Most of them are private chats and only 20 chats are group chats.

Imagine if I don't use the app for a day and at least 1 thousand chats are updated at least once. 

So, I can get only 1 thousand out of 10 thousand chats I am connected to. And I can check the updatedAt date of these 1 thousand chats only. Unfortunately, these 1k chats are not the same 1k chats which were updated, if I check the updatedAt on the client side. So I will have to check all the 10k chats to get the final results. I wouldn't like to download all these chats each time I enter the app. 

So, this is the first problem. How to get all the updated chats as fast as possible? Does this Join Table scenario really helps? Or is it better to check all the updated Chats on the server side using Many-to-Many Relations, as I do (and wait 10 sec)? 

II.
Refresh Dialogues while user is online

Second problem is how to get all the updated chats each time I go back to the main screen of the app called "My Chats". Me personally and most of my users go back and forth all the time, so I even had to set a timer on the app side, which checks if at least 30 sec passed before the "My Chats" section should be refreshed. If I wouldn't set that timer, the amount of requests to "Dialogues" would be not 4-5M per day, but up to 40-50M per day, which is impossible to host. Moreover, it really kills the phone`s battery.

PUSH-NOTIFICATIONS

A nice option you advised me to use Push notifications.

Sure, I do use them, but most of the clients, especially the active ones, don't let the app to receive push notifications. I personally don't allow any app, except iMail & iMessage, to send me any notifications, because they are annoying. But it also kills the process of refreshing chats if we rely on Push-notifications.

Ok. Let`s imagine that 100% of our users like push notifications. No problem here! But the problem is on the server side than. I send notifications in afterSave function, but the fact is that this function fails to often. Moreover, I heard from many users (and other Parse app founders) that push notifications are delayed too often. So, this is not the way to solve the problem. We cannot rely on push-notifications at all. Pushes are a kind of additional option which can make user`s life easier, but it`s not stable.

SOCKETS

Another way of receiving notifications which is 99% reliable is Sockets. We use PubNub. It processes 9.1M messages which are sent and received using our app. Yes, 9.1 million messages per day. They include 0.9M messages processed by Parse, and 8.2M messages received through sockets in private and group chats.

So, it would be great to refresh concrete chats each time they are updated, sending a socket to client and updating the "My Chats" section even without necessity to make queries to Parse, while user is online.

It really works with some of our beta-testers! Actually it works great, until you become an active client.

Let's recover again that active clients have thousands chats. So if a chat is updated, that particular active client should find that out in "My chats" section. Unfortunately, PubNub allows to subscribe to 100 channels maximum at once. So, in my scenario with my 10k private chats, it is absolutely impossible to find out which chats should I subscribe to, and which not. 100 is only 1% out of 10 000 chats for a given user.

Than I thought... Why wouldn't I try to get the most recent updates of each chat by receiving data through personal channel? I can create one channel per user and send each client the necessary info, right?

The problem is that most of the chats have more than 2 clients, so even if I try to send PubNub HttpRequests through Parse to these "personal" channels, I will end up with lots of fails and timeouts on the server side. There is no way I could handle an httprequest which should update 1000 personal channels at once. It works well only if we should send 1 httprequest per 1 updated dialogue.

I've contacted PubNub and described my problem. They said: "You can subscribe to group of channels for additional payment, it will allow you to subscribe a user up to 2000 channels at once".  Even if I do so, how do I get the updates of the rest 8k out of 10k channels, if I talk about my personal account?

Moreover, if I subscribe users to thousands channels, their phones will work non-stop receiving tons of data and their batteries will die fast.
------
To sum up, I couldn't find a better decision than to use these queries with Many-to-many relations.

But there must be another way. I hope you and your engineers can help me. Otherwise, in few months my average requests will take so long that all the shred will suffer.

**Comment by tunidev** [comment]
@halvini I am really interested for your feedback about using Parse server for a real time chat iOS app. So you use PubNub for messaging, is this a scalable and reliable solution ? any useful links please

**Comment by halvini** [comment]
PubNub is more scalable for chatting apps than Parse.
